### PR TITLE
refactor: Replace tracing with make_constraints_unique

### PR DESCRIPTION
**Pull request by lmondada**
Apologies for another large PR.

I've significantly reorganised and cleaned up the way the builder works:
- remove "graph tracing" which was hard to reason about. Instead, duplicate states before modifying them, then `try_merge_new_nodes` to reduce the number of states where possible.
- fix: previously the tracer could introduce cycles. Now checking before merging nodes that no cycles are created
- fix: handle matches in non-leaf positions of the constraint tree. It was possible before to create mutually exclusive trees that had constraint IDs in non-leaf nodes, but such cases were not handled properly. Now this is supported.
- Allow multiple transitions with same constraint during building, removing complicated recursive logic when adding new edges. The uniqueness of the transitions is restored one state at the time in the builder's `finish` step.
- Cleaner separation between automaton traversal and mutating API. The state automaton can no longer be modified outside of the builder.

### CODE REVIEW

File: src/automaton/builder.rs [hunk 1]
```diff
@@ -137,89 +181,212 @@ where
...
+                continue;
+            };
+            let siblings = self
+                .matcher
+                .incoming_transitions(first_child)
+                .map(|t| self.matcher.parent(t))
+                .unique()
+                .filter(|&n| n != state)
+                // We cannot merge nodes that are reachable from each other
+                .filter(|n| acyclic_checker.path_exists(n.0, state.0, &self.matcher.graph));
```
Comment (aborgna-q) [inline src/automaton/builder.rs:325]: This is not checking "each other" for all pairs.

Is it ok to merge `[n0, n1]` with `state` here?
```mermaid
flowchart
    u --> state
    u --> n0
    state
    n0 ---> n1
    state --> v
    n1 --> v
```

File: src/utils/subgraph.rs
Comment (aborgna-q) [inline src/utils/subgraph.rs:1]: Why not use [`NodeFiltered`](https://docs.rs/petgraph/latest/petgraph/visit/struct.NodeFiltered.html) ?

File: src/automaton/builder/modify.rs [hunk 2]
```diff
@@ -0,0 +1,318 @@
+use itertools::Itertools;
+
+use crate::automaton::{ConstraintAutomaton, State, StateID, Transition, TransitionID};
+use crate::PatternID;
+
+/// Methods for modifying the automaton
+///
+/// Exposed as a trait so that the automaton builder can reuse the default
+/// implementation but trace calls.
```
Comment (aborgna-q) [inline src/automaton/builder/modify.rs:9]: Outdated comment?

**Review by aborgna-q — APPROVED** [review]
Awesome!
Sorry this is not an in-depth review, but thanks for keeping me up-to-date.

File: src/automaton/builder/acyclic.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    /// Whether there is a path between `node1` and `node2` in the graph.
+    pub fn path_exists<G>(&self, node1: N, node2: N, graph: G) -> bool
+    where
+        G: IntoNeighborsDirected + IntoNodeIdentifiers + Visitable + GraphBase<NodeId = N>,
+    {
+        self.is_descendant(node1, node2, graph) || self.is_descendant(node2, node1, graph)
+    }
+
+    /// Merge all `nodes` into the first node in the slice.
+    pub fn merge_nodes(&mut self, nodes: impl Iterator<Item = N>) {
```
[the remark this answers is not in the document]
Comment (lmondada) [reply src/automaton/builder/acyclic.rs:71]: I've fixed the doc to clarify. Merging nodes removes all but the first one.

File: src/utils/subgraph.rs
Comment (lmondada) [reply src/utils/subgraph.rs:1]: ermm, that is a very good question. I must have not realised that I could compose `NodeFiltered` and `Reversed` so easily.

Thanks!!

**Comment by lmondada** [comment]
Thank you so much, very useful review!

**Pull request merged**